### ISSUE TITLE
Consider rename or alias for flux-broker command

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#643]
It was clear during a recent talk given by @trws that the `flux broker` command for starting a flux instance under another RM/PMI environment is confusing to users who may not even know what a broker is in this context.

Perhaps consider a more generic or user-friendly alias for `flux-broker`, or make `flux start` smart enough to determine if it is running in single-node-mode or under PMI environment so that either
```
srun -N8 flux start flux capacitor .,.
```
or
```
flux start flux capacitor
```

worked as expected. (first case is a PMI bootstrapped instance as a slurm job, second case is single node, single broker)

I'm not sure what you'd do with `srun -N8 flux start --size=4 COMMAND...` though ...

**Comment by garlick** [comment]
It would be nice if  we could make
```
srun -N8 flux initial-program
```
or even
```
flux initial-program
```

just work without contorting the behavior of the `flux` command driver beyond reason.  Maybe by special casing a missing `$FLUX_URI` and `--uri`?

**Comment by grondo** [comment]
Ha! I was going to suggest the same thing, but I didn't dare in fear of severe castigation.

I wasn't sure if there was a way to design the heuristic cleverly enough so that `flux` worked the same in and out of a `flux` instance. Additionally, to start capacitor would it have to be `srun -N8 flux flux capacitor`? In that case I thought `flux start flux capacitor` was actually more clear, and actually reads as grammar that does what you think it should. (same with `flux start initial-program`)

**Comment by garlick** [comment]
I was thinking just `srun -N8 flux capacitor` or `flux capacitor` would "just work" but I admit I haven't thought through how to get there!

**Comment by garlick** [comment]
Could we just special case execution under slurm by looking for `$SLURM_NNODES` or someting?  

So in `srun -N8 flux capacitor` flux execs `flux broker flux capacitor`?

**Comment by trws** [comment]
The question in that case is whether we want:

$ salloc -N 8
<job allocated>
$ flux capacitor

To run the equivalent of srun flux broker flux capacitor, because 
environment-variable wise wouldn't they look basically the same?

**Comment by grondo** [comment]
I don't feel like it is a good idea to start special casing resource managers, but maybe I'm just on a negativity spree today.

There might be a way to determine the difference between @trws's example above and `srun flux capacitor` from the environment, but it feels like this solution would be brittle for Slurm, and then we'd have to support a similar thing for `flux` itself, which I just have a feeling would be even more challenging.

**Comment by grondo** [comment]
> In `srun -N8 flux capacitor` `flux` execs `flux broker flux capacitor`

Would this require the `flux` cmddriver know a priori all subcommands which it could run? For example,  for `srun -N8 flux flugelhorn`, should the `flux` program exec `flux broker flugelhorn` or `flux broker flux flugelhorn`? That depends on if `flugelhorn` is an initial program or a flux subcommand...

**Comment by garlick** [comment]
Yeah... @grondo you're sensible as always.  Let's back up and reconsider the original proposal then.

**Comment by lipari** [comment]
So what about replacing `broker` with `run`?  The solution to @grondo's dilemma becomes `flux run flugelhorn` or `flux run flux flugelhorn`.  That would push the problem to disambiguating `flux run flugelhorn` from `flux wreckrun flugelhorn`.

**Comment by trws** [comment]
I was somewhat assuming wreck run would eventually become run, rather 
than having it be the instance launcher.  That does seem reasonable 
though.

**Comment by grondo** [comment]
> So what about replacing `broker` with `run`? 

I thought the same thing, but we do need to disambiguate between launching parallel work and launching a new instance under flux and other resource managers and environments. I do think @trws is correct, one we replace the wreck prototype the new thing should be `flux run`

However, I'm afraid I'm not fully grasping your suggestion @lipari. I think the proposal here is to enhance `flux start` so that it "starts an instance" whether in single node or under a resource manager or other multi-node launch.

However, I agree that under a flux instance `flux run [options] flux start program` does leave something to be desired. I could see us/users wanting a wrapper `flux start-instance [options].. program` (not suggesting that name in particular).

In fact under SLURM, users might like a similar wrapper `flux start-instance [OPTIONS] program` that calls `srun flux start/broker` under the covers as much as I don't like that idea in concept.

**Comment by garlick** [comment]
> make flux start smart enough to determine if it is running in single-node-mode or under PMI environment

This will probably be a messy heuristic involving RM-specific tests, but not a show stopper.  FYI SLURM's libpmi.so pretends its a size=1 job if not running in a SLURM allocation.

> I'm not sure what you'd do with srun -N8 flux start --size=4 COMMAND... though ...

Choose a default behavior when the situation is ambiguous, issue a warning, and provide a `--boot-method=pmi|local` option to override?

**Comment by trws** [comment]
Where my mind is, is that we could make two changes that together I think would ease this pretty nicely:

1  Have the run command do a check to see if there is a surrounding flux instance, and if not have it add a flux start with no arguments
2  Have flux start attempt to use PMI, and if that fails fall back on doing a local start

@garlick's point about having an option to force one method or the other makes sense to me as well, especially having it be extensible as it would be good to be able to add pmix or what have you as such things happen.

**Comment by garlick** [comment]
I can't see a problem with this bug @grondo should comment.

**Comment by lipari** [comment]
After reading the discussion above, throwing out ideas, and then thinking more about it, I'm inclined to reserve `flux run` as the eventual `flux wreckrun` replacement.  Instead I would expand `flux start` to include what `flux broker` does now.  This would provide an intuitive boundary between instantiating a flux instance and launching a program.

Perhaps we can consider adding -N <nodes> option to `flux start`.  This would allow us to better describe the distribution of brokers across:
* the allocated nodes when under an external RM allocation
* the parent flux instance

Still to do though is work out the "messy heuristic involving RM-specific tests" that @garlick foresees.

**Comment by garlick** [comment]
I think @trws means "run" as "wreckrun or future replacement"?  (Possibly we could just implement the logic the suggests in a new `flux-run` builtin that wraps wreckrun).  Or were you disagreeing with his suggestion 1 with this understanding?

**Comment by lipari** [comment]
I am retracting my original suggestion that `flux run` replace `flux broker`.  If @trws's item 1 was meant to build on that suggestion, then I am now disagreeing with it and suggesting `flux start` as its replacement.

**Comment by garlick** [comment]
Perhaps @trws can clarify what he meant.  This is starting to sound like dialog from _The Life of Brian_ :-)

**Comment by grondo** [comment]
> Where my mind is, is that we could make two changes that together I think would ease this pretty nicely:

> 1 Have the run command do a check to see if there is a surrounding flux instance, and if not have it add a flux start with no arguments

I think I see where you are going, but IIUC this scheme first requires `flux start` being able to replace or wrap `flux broker`, correct? When this is functional, your suggestion `1` is just a usability tweak so people don't have to type `srun flux start flux run...`. Does this also somehow improve usage like `srun flux start flux capacity [ARGS]...`?

> 2 Have flux start attempt to use PMI, and if that fails fall back on doing a local start

As @garlick mentioned, this seems more easily said than done, since SLURM's libpmi (in our default LD_LIBRARY_PATH) will appear to support PMI with n=1 no matter what.

**Comment by trws** [comment]
Now that you mention it @grondo, perhaps it doesn't.

What I meant to be proposing was that flux broker become part of flux start, and disambiguate by whether PMI is actually available or not.  Based on the point that it always supports PMI, but with a node count of one, I would amend that to checking for whether PMI is available, and if it is whether n is greater than 1.  Really, if n is 1, that's equivalent to having run on a single node anyway is it not?

In this context, flux run would be "wreckrun or eventual replacement thereto." That said, @grondo makes a good point that it's really not a win if it's just run, at least not as far as capacitor is concerned.  Perhaps it could be a function of the flux command itself as to whether it should imply a start or not based on a property of the command it is invoking?  Either way, it might not be so much of a win.  The major difference would be that "flux run ...", or perhaps more generally "flux <running command of some kind>", in a context without a flux instance would work rather than fail, and naturally do what srun would do in a context like inside a sized salloc.  That might be going a bit far for default behavior rather than a wrapper though.

**Comment by grondo** [comment]
> but with a node count of one, I would amend that to checking for whether PMI is available, and if it is whether n is greater than 1. Really, if n is 1, that's equivalent to having run on a single node anyway is it not

That's a pretty good insight and I'd say.

So based on that good idea, here's the steps I'd propose:
 1. Make `flux-start` enough of a PMI client to answer if PMI is available and `pmi_get_size() > 1`
 2. If 1. is true, then exec `/path/to/flux-broker`, otherwise fallback to old behavior
 3. If `--size` is provided force old behavior, add `--boot-method` option to disambiguate (e.g. I should have PMI_Size > 1, please die if not) as suggested by @garlick. 
 4. remove the path to `flux-broker` from `FLUX_EXEC_PATH` so that `broker` no longer appears as a subcommand. (Unless we want to keep that)

@garlick, does that sound reasonable? This seems to remove the need to handle RMs on a case by case basis, but I don't know if making `flux-start` a PMI peeker is a good or even feasible idea.

**Comment by garlick** [comment]
Sounds OK I think.

**Comment by garlick** [comment]
@grondo are you already working on this?  I just made a patch to fix #653, and maybe I should just keep going and finish this too?

**Comment by grondo** [comment]
I had started it but threw the code away, so feel free to complete!
